src/App.tsx

```diff
@@ -1,24 +1,190 @@
+    setTempTodo({
+      id: 0,
+      userId: todoService.USER_ID,
+      title: todoTitle,
+      completed: false,
+    });
+
+    setIsRequesting(true);
+
+    return todoService
+      .addTodo(todoTitle)
+      .then((newTodo) => {
+        setTodos((prevTodos) => [...prevTodos, newTodo]);
+      })
+      .catch(() => {
+        setErrorMessage(ErrorMessage.Add);
+      })
+      .finally(() => {
+        setIsRequesting(false);
+        setTempTodo(null);
+      });
+  };
+
+  const handleDeleteTodo = (todoId: number) => {
+    setProcessingTodoIds((prevTodoIds) => [...prevTodoIds, todoId]);
+
+    todoService
+      .deleteTodo(todoId)
+      .then((() => {
+        setTodos((prevTodos) => prevTodos.filter(todo => todo.id !== todoId));
+      }))
+      .catch(() => {
+        setErrorMessage(ErrorMessage.Delete);
+      })
+      .finally(() => {
+        setProcessingTodoIds(
+          (prevTodoIds) => prevTodoIds.filter(id => id !== todoId),
+        );
+      });
+  };
+
+  const handleCompletedChange = (todo: Todo) => {
+    setProcessingTodoIds((prevTodoIds) => [...prevTodoIds, todo.id]);
+
+    todoService
+      .updateTodo({
+        ...todo,
+        completed: !todo.completed,
+      })
+      .then(updatedTodo => {
+        setTodos(prevState => prevState.map(currentTodo => (
+          currentTodo.id !== updatedTodo.id
+            ? currentTodo
+            : updatedTodo
+        )));
+      })
+      .catch(() => {
+        setErrorMessage(ErrorMessage.Update);
+      })
+      .finally(() => {
+        setProcessingTodoIds(
+          (prevTodoIds) => prevTodoIds.filter(id => id !== todo.id),
+        );
+      });
+  };
+
+  const handleDeleteAllCompletedTodos = () => {
+    todos
+      .filter((todo) => todo.completed)
+      .forEach((todo) => handleDeleteTodo(todo.id));
+  };
+
+  const handleUpdateTodo = (todo: Todo, newTodoTitle: string) => {
+    setProcessingTodoIds((prevTodoIds) => [...prevTodoIds, todo.id]);
+
+    return todoService
+      .updateTodo({
+        ...todo,
+        title: newTodoTitle,
+      })
+      .then(updatedTodo => {
+        setTodos(prevState => prevState.map(currentTodo => (
+          currentTodo.id !== updatedTodo.id
+            ? currentTodo
+            : updatedTodo
+        )));
+      })
+      .catch(() => {
+        setErrorMessage(ErrorMessage.Update);
+      })
+      .finally(() => {
+        setProcessingTodoIds(
+          (prevTodoIds) => prevTodoIds.filter(id => id !== todo.id),
+        );
+      });
+  };
+
+  const todosCounter = useMemo(() => {
+    return todos.filter(todo => !todo.completed).length;
+  }, [todos]);
+
+  const isAllCompleted = useMemo(() => {
+    return todos.every(todo => todo.completed);
+  }, [todos]);
+
+  const hasCompletedTodo = todos.some(({ completed }) => completed);
+
+  const handleToggleButton = () => {
+    todos
+      .filter((todo) => todo.completed === isAllCompleted)
+      .forEach((todo) => handleCompletedChange(todo));
```

Comment:
It is not necessary to go around the array twice, it is enough to move the filtering condition to the middle of the forEach and go through the array once instead of twice